### ISSUE TITLE
jline3 line-wrapping buglet and fix

### ISSUE THREAD

**Issue by PerBothner** [jline/jline2#256]
Copy-paste from a wrapped line looses the wrap information.
I'm pretty sure this is because noWrapAtEol is incorrectly set in the Display constructor.
The given initialization is true when ansi-style wrap _is_ supported, so it should actually be:

```
    this.noWrapAtEol = ! (terminal.getBooleanCapability(Capability.auto_right_margin)
                          && terminal.getBooleanCapability(Capability.eat_newline_glitch));
```

No pull request, since I'm not yet familiar with this aspect of Git, but this one is pretty trivial.

**Comment by PerBothner** [comment]
See pull request https://github.com/jline/jline2/pull/258

**Comment by gnodet** [comment]
This needs to be reinvestigated, as I had to revert the change to fix the display.

When ```noWrapAtEol``` is true, it means the cursor does not automatically go to the beginning of the next line, and thus needs special handling.

Can you give a way to reproduce your copy/paste problem ? Does it happen when pasting a string with a new line in it ?  On which OS/Terminal ?

**Comment by PerBothner** [comment]
In a jline3 application on xterm-style terminals (tested on xterm and Gnome Terminal) type a long input line, one that is longer than the screen width, so it wraps around.
Then select and copy the input and paste into (say) emacs.
With the patch, the long wrapped input line becomes a long wrapped input line in emacs.
Without the patch, it becomes multiple lines.
Bash w/readline handles this fine.

**Comment by gnodet** [comment]
Ok, I do understand the use case.  However, this is definitely not the correct way to handler that I think.
Currently, the Display class does not support lines longer than the terminal.  All code using it do the line splitting before giving the list of "visual" lines to display to the Display instance (see less, nano, or even the line reader).
Support for the use case described would require a bigger change imho.

**Comment by PerBothner** [comment]
The implemented approach seems relatively logical and clean (based on a quick look), and it seems to work (barring some glitches).  The support for this use-case precedes my patch - I'm guessing the logic was imported from jline2, but there was some mistake in initializing noWrapAtEol.

The whole point of noWrapAtEol is to support proper copy-and-paste of wrapped lines - otherwise I see no reason for it.

I think we should fix whatever specific display bug we're experiencing, rather than rip the support out.

**Comment by gnodet** [comment]
No the noWrapAtEol is not related to copy/paste at all.
The only purpose of this flag is to be able to write something on the right edge of the terminal correctly.  This flag is actually to fix display bugs, it's definitely not for supporting wrapping, which is not supported at all.

You can look at the jline 1 / jline 2 history and you'll see this flag history.

The flag is available to discriminate the way some terminal works.  When writing the last char on a given line, some terminals to automatically move the cursor to the beginning of the next line.  Some don't.   If you look at the code, the flag is only used to know if the cursor is correctly positionned and if not, to position it correctly.

So I'm all for implementing line wrapping correctly, but that's definitely not the way to do it.

Currently, the code for the Display class is supposed to receive only lines that are at most the size of the terminals (i.e. no support for wrapping).

**Comment by PerBothner** [comment]
The noWrapAtEol variable conflates two things that should be handled separately:

1. no wrapping at eol;
2. delayed wrapping at eol - the wrapping happens after the next character.

They should be handled differently, controlled by two variables, maybe noWrapAtEol and delayedWrapAtEol. The existing noWrapAtEol handling seems fine.

A simple but inefficient fix for delayedWrapAtEol is to print the next character (the one that will be on the following line) followed by a backspace to get back to column 0.  Something like:

            if (!cursorOk && cursorPos == curCol + columns) {
                if (noWrapAtEol
                    || (delayedWrapAtEol && isLastLine)) {
                    terminal.puts(Capability.carriage_return); // CR / not newline.
                    cursorPos = curCol;
                    cursorOk = true;
                } else if (delayedWrapAtEol) {
                    rawPrint(newLines.get(lineIndex+1).substring(0, 1));
                    terminal.puts(Capability.cursor_left);
                    cursorOk = true;
                }
            }

The disadvantage is that very frequently you will either write the next character (the one you just write and backspaced over), or move the cursor further down.

A better but more complex fix would use an extra field cursorAtEol:

            if (!cursorOk && cursorPos == curCol + columns) {
                if (noWrapAtEol) {
                    terminal.puts(Capability.carriage_return); // CR / not newline.
                    cursorPos = curCol;
                } else if (delayedWrapAtEol) {
                    cursorAtEol = true;
                }
                cursorOk = true;
            }

This variable is a companion to cursorPos: Both must be considered by moveVisualCursorTo.  

Most calls to moveVisualCursorTo are immediately followed by a call to rawPrint - and most commonly the position moved to is the current position.  So I suggest combining those to call to a single function:

    moveCursorAndPrint(int position, AttributedString text, AttributedString oldLine)

oldLine is the existing string from oldLines.get(lineIndex+1), or null in the case of the last line.

When cursorAtEol is false, just do moveVisualCursorTo followed by rawPrint (when text != null).

When cursorAtEol && position==cursorPos && text != null; just write text.

Otherwise, if cursorAtEol write the first character of oldLine, then backSpace, then continue as if cursorAtEol is false.  (If oldLine is false, use space.)

The cases where moveVisualCursorTo is not followed by a rawPrint use moveCursorAndWrite(position, null, oldLine).

**Comment by gnodet** [comment]
One thing which is absent from your logic is the recognition of wether a newline should be printed or not.
Where is the difference between 5 lines of 80 characters (because a justified table is displayed for example) and a single line of 5*80 chars is made ?  I think in the first case, you want to keep the current behavior, while in the second, you'd want to leverage the wrapping of the terminal as you suggest.

**Comment by PerBothner** [comment]
You're right.  One idea is the change columnSplitLength so newlines are included in each line:

            if (cp == '\n') {
                strings.add(subSequence(beg, cur+1);

Perhaps depending on a boolean parameter:
 
           if (cp == '\n') {
                strings.add(subSequence(beg, includeNewlines ? cur+1 : cur);

This has the neat property that concatenating all the strings in the result gives you the original string.

It is now easy to distinguish the cases you mention.  Of course that requires even more changes to the update logic.

(This idea seems familiar - I may have proposed it before - or maybe I just thought of it.)

**Comment by gnodet** [comment]
I'm not sure I actually get your proposal about ```moveCursorAndPrint```.
On the 12 calls to ```moveCursorAndPrint```, only 3 are immediately followed by a call to ```rawPrint```.  Usually, there's a call to ```insertLines/deleteLines/insertCharts/deleteChars``` between those 2 calls.  The whole goal of the Display class is to minimize the operations made to draw the changes, so when the user enters a char or backspace in the middle of the line, the whole line is not redrawn completely, that's why we compute the diffs between the previous and new state and use the diffs to compute minimal changes.

**Comment by PerBothner** [comment]
"On the 12 calls to moveCursorAndPrint, only 3 are immediately followed by a call to rawPrint".

Some more calls appear if you move the moveVisualCursorTo calls inside an if. For example:

        while (lineIndex < Math.max(oldLines.size(), newLines.size())) {
            if (lineIndex < oldLines.size()) {
                moveCursorAndPrint(currentPos, null, null);
                 ...
            } else {
                moveCursorAndPrint(currentPos, newLines.get(lineIndex), null);
                ...
            }

The goal of combining moveVisualCursorTo and rawPrint is a minor optimization. When the optimization does not apply (or if it's not worth the extra complication), it is fine to resolve the cursorAtEol case by printing the (existing or new) first character of the next line, optionally followed by a backspace, depending on where you need to end up.

**Comment by gnodet** [comment]
I was thinking about releasing a JLine 3.0.0 before this thread.  Do you plan to work on something in the short term, or should I start a 3.0.0 release and we can defer and plan this enhancement for 3.1.0.

**Comment by gnodet** [comment]
Also, I'll create an issue in the jline3 project, where it will make more sense.

**Comment by PerBothner** [comment]
No, I'm not planning on working on this in the short term. I need to get my own release out.

**Comment by PerBothner** [comment]
I created a pull request: [Support correct line wrapping (fixes issue #25) #34](https://github.com/jline/jline3/pull/34).

I've done some (but not all) of the testing using DomTerm, which has the advantage that it displays continuation (wrap) markers, Emacs-style.  The easiest way to see it in action  is to try a Kawa snapshot:  ftp://ftp.gnu.org/pub/gnu/kawa/kawa-latest.zip .. Just run:

    kawa-2.1.91/bin/kawa -w

The -w should pop up a window with "Kawa (DomTerm)" in the title bar.
Kawa sets the DELAY_LINE_WRAP option (if running under DomTerm).
